### PR TITLE
JAVA-2265: Configure NullSavingStrategy at the DAO level

### PR DESCRIPTION
**Pull request by tomekl007**
…or Update annotation

### CODE REVIEW

File: mapper-processor/src/main/java/com/datastax/oss/driver/internal/mapper/processor/dao/DaoMethodGenerator.java [hunk 1]
```diff
@@ -214,6 +218,36 @@ protected void maybeAddTimestamp(String timestamp, MethodSpec.Builder methodBuil
     }
   }
 
+  protected boolean nullSavingStrategyExplicitlySet(
+      ExecutableElement methodElement, String annotationCanonicalName) {
```
Comment (tomekl007) [inline mapper-processor/src/main/java/com/datastax/oss/driver/internal/mapper/processor/dao/DaoMethodGenerator.java]: This is code responsible for finding out if the `nullSavingStrategy` value was explicitly set on the `@Update` annotation using `AnnotationMirror`. It looks a bit complex, @olim7t do you know if there is a way to create `ExecutableElement` using java-poet for the purpose of this test?

File: mapper-runtime/src/main/java/com/datastax/oss/driver/api/mapper/annotations/DefaultNullSavingStrategy.java [hunk 2]
```diff
@@ -0,0 +1,46 @@
...
+
+import com.datastax.oss.driver.api.mapper.entity.saving.NullSavingStrategy;
+import java.lang.annotation.ElementType;
+import java.lang.annotation.Retention;
+import java.lang.annotation.RetentionPolicy;
+import java.lang.annotation.Target;
+
+@Target(ElementType.TYPE)
+@Retention(RetentionPolicy.CLASS)
+public @interface DefaultNullSavingStrategy {
```
Comment (tomekl007) [inline mapper-runtime/src/main/java/com/datastax/oss/driver/api/mapper/annotations/DefaultNullSavingStrategy.java:57]: I think If we choose to add new annotation `DefaultNullSavingStrategy` this annotation should have `nullSavingStrategy` without default. If someone is adding that at the `@Dao` level it means that he must specify `nullSavingStrategy` that will apply to inherited strategy for methods: `@Update`, `@Insert`, `@SetEntity`, `@Query`.

**Review by tomekl007 — COMMENTED** [review]
Implemented only for `@Update` and waiting for feedback.
`@SetEntity`, `@Insert` and `Query` level handing todo (it is fairly simple: code from line https://github.com/datastax/java-driver/pull/1246/files#diff-25ba2cdaf8c214d0d4a6f05f1c7922acR129 to https://github.com/datastax/java-driver/pull/1246/files#diff-25ba2cdaf8c214d0d4a6f05f1c7922acR135)

File: mapper-runtime/src/main/java/com/datastax/oss/driver/api/mapper/annotations/DefaultNullSavingStrategy.java (hunk 2 above)
Comment (olim7t) [reply mapper-runtime/src/main/java/com/datastax/oss/driver/api/mapper/annotations/DefaultNullSavingStrategy.java:57]: I agree with no default.

File: mapper-processor/src/main/java/com/datastax/oss/driver/internal/mapper/processor/dao/DaoMethodGenerator.java (hunk 1 above)
Comment (olim7t) [reply mapper-processor/src/main/java/com/datastax/oss/driver/internal/mapper/processor/dao/DaoMethodGenerator.java]: JavaPoet only generates strings. The only way to create an `ExecutableElement` for a test is either through mocking or with a full compiler execution I think.

File: mapper-processor/src/main/java/com/datastax/oss/driver/internal/mapper/processor/dao/DaoMethodGenerator.java [hunk 3]
```diff
@@ -214,6 +218,36 @@ protected void maybeAddTimestamp(String timestamp, MethodSpec.Builder methodBuil
...
+        .filter(
+            v -> {
+              // Find only for Update annotation
+              DeclaredType annotationType = v.getAnnotationType();
+              if (annotationType instanceof Type.ClassType) {
+                return ((Type.ClassType) annotationType)
+                    .tsym
+                    .getQualifiedName()
+                    .toString()
+                    .equals(annotationCanonicalName);
```
Comment (olim7t) [inline mapper-processor/src/main/java/com/datastax/oss/driver/internal/mapper/processor/dao/DaoMethodGenerator.java]: You can compare the type with a `Class` directly:
```java
context.getClassUtils().isSame(getAnnotationType(), Update.class)
```

File: mapper-processor/src/main/java/com/datastax/oss/driver/internal/mapper/processor/dao/DaoMethodGenerator.java [hunk 4]
```diff
@@ -214,6 +218,36 @@ protected void maybeAddTimestamp(String timestamp, MethodSpec.Builder methodBuil
...
+              } else {
+                return false;
+              }
+            })
+        .flatMap(v -> v.getElementValues().values().stream())
+        .filter(v -> v instanceof Attribute.Enum)
```
Comment (olim7t) [inline mapper-processor/src/main/java/com/datastax/oss/driver/internal/mapper/processor/dao/DaoMethodGenerator.java]: From `getElementValues()` javadocs:
> Only those elements with values explicitly present in the annotation are included, not those that are implicitly assuming their default values.

So we just need to iterate `getElementValues().keySet()` and find out if there is an element such that `key.getSimpleName().contentEquals("nullSavingStrategy")`.

File: mapper-processor/src/main/java/com/datastax/oss/driver/internal/mapper/processor/dao/DaoMethodGenerator.java [hunk 5]
```diff
@@ -214,6 +218,36 @@ protected void maybeAddTimestamp(String timestamp, MethodSpec.Builder methodBuil
     }
   }
 
+  protected boolean nullSavingStrategyExplicitlySet(
+      ExecutableElement methodElement, String annotationCanonicalName) {
+    return !methodElement
+        .getAnnotationMirrors()
+        .stream()
```
Comment (olim7t) [inline mapper-processor/src/main/java/com/datastax/oss/driver/internal/mapper/processor/dao/DaoMethodGenerator.java]: I think the Stream API really makes things worse here (and to be honest that's true most of the time -- that's why it's discouraged in the contribution guidelines).

There's an imperative version for a similar case [here](https://github.com/datastax/java-driver/blob/java2078/mapper-processor/src/main/java/com/datastax/oss/driver/internal/mapper/processor/dao/DaoQueryProviderMethodGenerator.java#L84). It's not pretty either, but I find it way more readable.

Comment (tomekl007) [reply mapper-processor/src/main/java/com/datastax/oss/driver/internal/mapper/processor/dao/DaoMethodGenerator.java]: I've changed (https://github.com/datastax/java-driver/pull/1246/commits/75a978baf7118914f2907e47e89a46ead8e7561f) it to imperative way but I am not sure if this is more readable that stream versions after simplification:
```
  return methodElement
        .getAnnotationMirrors()
        .stream()
        .filter(v -> classUtils.isSame(v.getAnnotationType(), javaClass))
        .flatMap(v -> v.getElementValues().keySet().stream())
        .anyMatch(v -> v.getSimpleName().contentEquals("nullSavingStrategy"));
```

**Comment by tolbertam** [comment]
Sorry for delay in review, this will be my first priority tomorrow :+1:

File: mapper-processor/src/main/java/com/datastax/oss/driver/internal/mapper/processor/dao/DaoImplementationGenerator.java [hunk 6]
```diff
@@ -320,56 +328,59 @@ private void generateProtocolVersionCheck(MethodSpec.Builder builder) {
...
+  private void addVersionCheckToBuilder(MethodSpec.Builder builder) {
+    builder
+        .beginControlFlow(
+            "if (context.getSession().getContext().getProtocolVersion().getCode() <= $T.Version.V3)",
+            ProtocolConstants.class)
+        .addStatement(
+            "throw new IllegalArgumentException($S)",
+            "You cannot use NullSavingStrategy.DO_NOT_SET for protocol version V3.")
+        .endControlFlow()
+        .build();
```
Comment (olim7t) [inline mapper-processor/src/main/java/com/datastax/oss/driver/internal/mapper/processor/dao/DaoImplementationGenerator.java]: You could move this to a static method on `DaoBase`, so that you only have to generate the method call.

On the other hand the code is still relatively simple so it's not a big deal to generate it. Your call.

File: mapper-runtime/src/main/java/com/datastax/oss/driver/api/mapper/annotations/DefaultNullSavingStrategy.java [hunk 7]
```diff
@@ -0,0 +1,46 @@
...
+   * of: {@link Update#nullSavingStrategy()}, {@link Insert#nullSavingStrategy()}, {@link
+   * SetEntity#nullSavingStrategy()} {@link Query#nullSavingStrategy()} explicitly the {@code
+   * DefaultNullSavingStrategy#nullSavingStrategy()} will be taken.
+   *
+   * <p>If you do not specify {@code DefaultNullSavingStrategy} at the {@link Dao} level the default
+   * or explicitly user set values from one of: {@link Update#nullSavingStrategy()}, {@link
+   * Insert#nullSavingStrategy()}, {@link SetEntity#nullSavingStrategy()} {@link
+   * Query#nullSavingStrategy()} will be taken
+   */
+  NullSavingStrategy nullSavingStrategy();
```
Comment (olim7t) [inline mapper-runtime/src/main/java/com/datastax/oss/driver/api/mapper/annotations/DefaultNullSavingStrategy.java]: I think we don't need a dedicated parameter name here, since it's the only one and it's pretty obvious from the name of the annotation. So I would call this method `value()` so that it can be written directly like this:

```java
@DefaultNullSavingStrategy(SET_TO_NULL)
```

File: manual/mapper/daos/defaultnullsavingstrategy/README.md [hunk 8]
```diff
@@ -0,0 +1,33 @@
+## DefaultNullSavingStrategy
+
+If you don't specify `nullSavingStrategy` parameter on the method level you can suppress 
+defaults by using `@DefaultNullSavingStrategy` annotation.
```
Comment (olim7t) [inline manual/mapper/daos/defaultnullsavingstrategy/README.md]: Can you make `@DefaultNullSavingStrategy` a link to the Javadocs?

File: manual/mapper/daos/null_saving/README.md [hunk 9]
```diff
@@ -39,6 +39,11 @@ Two strategies are available:
     
 If you don't specify an explicit strategy for a method, the mapper defaults to `SET_TO_NULL`.
 
+## Default Dao level null saving strategy
+
+If you don't specify `nullSavingStrategy` parameter on the method level you can suppress defaults by using
+`@DefaultNullSavingStrategy` annotation. See [@DefaultNullSavingStrategy](../defaultnullsavingstrategy/) for more details.
```
Comment (olim7t) [inline manual/mapper/daos/null_saving/README.md]: I don't think a separate page is necessary here, I would inline it in this section.
It fits in nicely since it relates to the main topic.

File: mapper-processor/src/main/java/com/datastax/oss/driver/internal/mapper/processor/dao/DaoImplementationGenerator.java [hunk 10]
```diff
@@ -71,11 +69,13 @@
   private final Map<ClassName, String> entityHelperFields = new LinkedHashMap<>();
   private final List<GeneratedPreparedStatement> preparedStatements = new ArrayList<>();
   private final List<GeneratedQueryProvider> queryProviders = new ArrayList<>();
+  private final NullSavingStrategyValidation nullSavingStrategyValidation;
 
   public DaoImplementationGenerator(TypeElement interfaceElement, ProcessorContext context) {
     super(context);
     this.interfaceElement = interfaceElement;
     implementationName = GeneratedNames.daoImplementation(interfaceElement);
+    nullSavingStrategyValidation = new NullSavingStrategyValidation(context);
```
Comment (olim7t) [inline mapper-processor/src/main/java/com/datastax/oss/driver/internal/mapper/processor/dao/DaoImplementationGenerator.java:77]: 👍 Good idea to extract the boilerplate into a separate component

File: mapper-processor/src/main/java/com/datastax/oss/driver/internal/mapper/processor/dao/DaoInsertMethodGenerator.java [hunk 11]
```diff
@@ -127,8 +129,21 @@ public DaoInsertMethodGenerator(
 
     List<? extends VariableElement> parameters = methodElement.getParameters();
     String entityParameterName = parameters.get(0).getSimpleName().toString();
-    NullSavingStrategy nullSavingStrategy =
-        methodElement.getAnnotation(Insert.class).nullSavingStrategy();
+
+    Insert annotation = methodElement.getAnnotation(Insert.class);
+    NullSavingStrategy nullSavingStrategy;
+
+    Optional<NullSavingStrategy> defaultNullSavingStrategy = enclosingClass.getNullSavingStrategy();
```
Comment (olim7t) [inline mapper-processor/src/main/java/com/datastax/oss/driver/internal/mapper/processor/dao/DaoInsertMethodGenerator.java]: Maybe calling this `daoNullSavingStrategy` would make the code below a bit easier to read?

File: mapper-processor/src/main/java/com/datastax/oss/driver/internal/mapper/processor/dao/NullSavingStrategyValidation.java [hunk 12]
```diff
@@ -0,0 +1,246 @@
...
+import com.datastax.oss.driver.internal.mapper.processor.ProcessorContext;
+import com.datastax.oss.driver.internal.mapper.processor.util.Classes;
+import java.util.List;
+import java.util.Optional;
+import javax.lang.model.element.AnnotationMirror;
+import javax.lang.model.element.Element;
+import javax.lang.model.element.ExecutableElement;
+import javax.lang.model.element.TypeElement;
+
+class NullSavingStrategyValidation {
```
Comment (olim7t) [inline mapper-processor/src/main/java/com/datastax/oss/driver/internal/mapper/processor/dao/NullSavingStrategyValidation.java]: We might want to make this class (and its two package-private methods) public.
We're not committing to the API because it's in an internal package, and it could come in handy if we extend the generators or create new ones from the DSE mapper.

File: mapper-processor/src/main/java/com/datastax/oss/driver/internal/mapper/processor/dao/NullSavingStrategyValidation.java [hunk 13]
```diff
@@ -0,0 +1,246 @@
...
+      return true;
+      // if DAO level DO_NOT_SET check if all underlying override it explicitly to SET_TO_NULL
+    } else if (hasDaoLevelDoNotSetStatement(daoLevelInterface)
+        && !hasSetToNullStatementSetExplicitlyForAll) {
+      return true;
+      // if DAO level annotation do not present, check method level strategy
+      // (including the default one)
+    } else {
+      return hasDaoLevelUnset(daoLevelInterface) && hasDoNotSetStatement;
+    }
```
Comment (olim7t) [inline mapper-processor/src/main/java/com/datastax/oss/driver/internal/mapper/processor/dao/NullSavingStrategyValidation.java:98]: This is a bit hard to follow, but I think it's more because of the inherent complexity, I'm not sure I could do better :)

Maybe a few naming suggestions (as you see fit):
```
hasDoNotSetStatement => anyMethodHasOrDefaultsToDoNotSet
hasDoNotSetStatementSetExplicitly => anyMethodHasDoNotSetExplicitly
hasSetToNullStatementSetExplicitlyForAll => allMethodsHaveSetToNull (no need to say "explicitly" here, SET_TO_NULL is always explicit)

hasDaoLevelSetToNullStatement => daoHasSetToNull
hasDaoLevelDoNotSetStatement => daoHasDoNotSet
hasDaoLevelUnset => daoIsNotAnnotated
```

File: mapper-runtime/src/main/java/com/datastax/oss/driver/api/mapper/annotations/DefaultNullSavingStrategy.java [hunk 14]
```diff
@@ -0,0 +1,46 @@
...
+   * <p>If you specify {@link DefaultNullSavingStrategy} at the {@link Dao} level and do not set one
+   * of: {@link Update#nullSavingStrategy()}, {@link Insert#nullSavingStrategy()}, {@link
+   * SetEntity#nullSavingStrategy()} {@link Query#nullSavingStrategy()} explicitly the {@code
+   * DefaultNullSavingStrategy#nullSavingStrategy()} will be taken.
+   *
+   * <p>If you do not specify {@code DefaultNullSavingStrategy} at the {@link Dao} level the default
+   * or explicitly user set values from one of: {@link Update#nullSavingStrategy()}, {@link
+   * Insert#nullSavingStrategy()}, {@link SetEntity#nullSavingStrategy()} {@link
+   * Query#nullSavingStrategy()} will be taken
+   */
```
Comment (olim7t) [inline mapper-runtime/src/main/java/com/datastax/oss/driver/api/mapper/annotations/DefaultNullSavingStrategy.java]: I would move this as the main javadoc for the annotation. Also a few suggestions to reorganize the explanations:
```java
/**
 * Annotates a {@link Dao} interface to define a default {@link NullSavingStrategy}, that will apply
 * to all methods that don't explicitly declare one.
 *
 * <p>For example, given this interface:
 *
 * <pre>
 * &#64;Dao
 * &#64;DefaultNullSavingStrategy(SET_TO_NULL)
 * public interface ProductDao {
 *
 *   &#64;Insert
 *   void insert(Product product);
 *
 *   &#64;Update(nullSavingStrategy = DO_NOT_SET)
 *   void update(Product product);
 * }
 * </pre>
 *
 * <ul>
 *   <li>{@code insert(Product)} will use {@link NullSavingStrategy#SET_TO_NULL SET_TO_NULL}
 *       (inherited from the DAO's default).
 *   <li>{@code update(Product)} will use {@link NullSavingStrategy#DO_NOT_SET DO_NOT_SET}.
 * </ul>
 *
 * If the DAO interface isn't annotated with {@link DefaultNullSavingStrategy}, any method that does
 * not declare its own value defaults to {@link NullSavingStrategy#DO_NOT_SET DO_NOT_SET}.
 *
 * <p>Note that null saving strategies are only relevant for {@link Update}, {@link Insert}, {@link
 * Query} and {@link SetEntity} methods.
 */
```

**Review by olim7t — CHANGES_REQUESTED** [review]

File: mapper-runtime/src/main/java/com/datastax/oss/driver/api/mapper/annotations/DefaultNullSavingStrategy.java (hunk 2 above)
Comment (tolbertam) [inline mapper-runtime/src/main/java/com/datastax/oss/driver/api/mapper/annotations/DefaultNullSavingStrategy.java:57]: What do you think about naming this simply `@NullSavingStrategy`?  Default seems kind of implicit, as to me it is intuitive to expect that if you were to also define `nullSavingStrategy` at the method-annotation level that it would override the interface level annotation.

File: mapper-processor/src/main/java/com/datastax/oss/driver/internal/mapper/processor/dao/NullSavingStrategyValidation.java [hunk 15]
```diff
@@ -0,0 +1,246 @@
...
+      return annotation.nullSavingStrategy() == NullSavingStrategy.SET_TO_NULL;
+    }
+    return false;
+  }
+
+  private boolean hasDaoLevelUnset(TypeElement interfaceElement) {
+    return interfaceElement.getAnnotation(DefaultNullSavingStrategy.class) == null;
+  }
+
+  private boolean queryHasDoNotSet(ExecutableElement v, boolean explicitSet) {
```
Comment (tolbertam) [inline mapper-processor/src/main/java/com/datastax/oss/driver/internal/mapper/processor/dao/NullSavingStrategyValidation.java:123]: There is a lot of boilerplate here between `queryHasDoNotSet`, `setEntityHasDoNotSet`, `insertHasDoNotSet` and `updateHasDoNotSet` that could probably be factored into a common method, for example you could do something like this:

```java
  private <A extends Annotation> boolean hasDoNotSet(Class<A> clazz,
                              Function<A, NullSavingStrategy> extractor,
                              ExecutableElement v,
                              boolean explicitSet) {
    A annotation = v.getAnnotation(clazz);
    if (annotation != null) {
      NullSavingStrategy strategy = extractor.apply(annotation);
      if (explicitSet) {
        return strategy == NullSavingStrategy.DO_NOT_SET
          && nullSavingStrategyExplicitlySet(v, clazz);
      } else {
        return strategy == NullSavingStrategy.DO_NOT_SET;
      }
    }
    return false;
  }

  private boolean queryHasDoNotSet(ExecutableElement v, boolean explicitSet) {
    return hasDoNotSet(Query.class, Query::nullSavingStrategy, v, explicitSet);
  }
```

File: mapper-processor/src/main/java/com/datastax/oss/driver/internal/mapper/processor/dao/NullSavingStrategyValidation.java [hunk 16]
```diff
@@ -0,0 +1,246 @@
...
+            && nullSavingStrategyExplicitlySet(v, Update.class);
+      } else {
+        return annotation.nullSavingStrategy() == NullSavingStrategy.DO_NOT_SET;
+      }
+    } else {
+      return false;
+    }
+  }
+
+  private boolean queryHasSetToNullExplicitly(ExecutableElement v) {
```
Comment (tolbertam) [inline mapper-processor/src/main/java/com/datastax/oss/driver/internal/mapper/processor/dao/NullSavingStrategyValidation.java:157]: same here, these `HasSetToNullExplicitly` methods could be factored into a common method.

File: mapper-processor/src/main/java/com/datastax/oss/driver/internal/mapper/processor/dao/NullSavingStrategyValidation.java [hunk 17]
```diff
@@ -0,0 +1,246 @@
...
+  }
+
+  private boolean isOperationWithNullSavingStrategy(ExecutableElement v) {
+    return v.getAnnotation(Update.class) != null
+        || v.getAnnotation(Insert.class) != null
+        || v.getAnnotation(SetEntity.class) != null
+        || v.getAnnotation(Query.class) != null;
+  }
+
+  boolean nullSavingStrategyExplicitlySet(Element methodElement, Class<?> javaClass) {
```
Comment (tolbertam) [inline mapper-processor/src/main/java/com/datastax/oss/driver/internal/mapper/processor/dao/NullSavingStrategyValidation.java]: If we were to remove `default NullSavingStrategy.DO_NOT_SET` from all annotations so they are null instead couldn't this be more effective as all you would have to do is null check?

File: manual/mapper/daos/defaultnullsavingstrategy/README.md [hunk 18]
```diff
@@ -0,0 +1,33 @@
+## DefaultNullSavingStrategy
+
+If you don't specify `nullSavingStrategy` parameter on the method level you can suppress
```
Comment (tolbertam) [inline manual/mapper/daos/defaultnullsavingstrategy/README.md]: instead of `you can suppress defaults`, I would recommend `you can adjust the default behavior`

File: manual/mapper/daos/defaultnullsavingstrategy/README.md [hunk 19]
```diff
@@ -0,0 +1,33 @@
...
+
+Strategies when using `@DefaultNullSavingStrategy` with method level `nullSavingStrategy` on annotations: 
+[@Insert](../insert/), [@Query](../query/), [@SetEntity](../setentity/) or
+[@Update](../update/):
+
+* If you specify `@DefaultNullSavingStrategy` at the [@Dao] level and set `nullSavingStrategy` parameter on one of the 
+  [@Update](../update/), [@Insert](../insert/), [@Query](../query/) or [@SetEntity](../setentity./)
+  explicitly, the method level annotation value will be taken.
+
+* If you specify `@DefaultNullSavingStrategy` at the [@Dao] level and do not set `nullSavingStrategy` parameter on one of the
```
Comment (tolbertam) [inline manual/mapper/daos/defaultnullsavingstrategy/README.md]: These next two paragraphs don't add much as I think it should be apparent that if you only use one of the ways to configure null saving, the one that is used is used.

File: manual/mapper/daos/defaultnullsavingstrategy/README.md [hunk 20]
```diff
@@ -0,0 +1,33 @@
...
+```java
+@Dao
+@DefaultNullSavingStrategy(nullSavingStrategy = NullSavingStrategy.SET_TO_NULL)
+public interface DefaultNullStrategyDao {
+  // dao methods
+}
+
+```
+
+Strategies when using `@DefaultNullSavingStrategy` with method level `nullSavingStrategy` on annotations:
```
Comment (tolbertam) [inline manual/mapper/daos/defaultnullsavingstrategy/README.md]: If you agree and remove the two paragraphs below, you don't need this first line either.

File: mapper-processor/src/main/java/com/datastax/oss/driver/internal/mapper/processor/dao/DaoInsertMethodGenerator.java [hunk 21]
```diff
@@ -127,8 +129,21 @@ public DaoInsertMethodGenerator(
 
     List<? extends VariableElement> parameters = methodElement.getParameters();
     String entityParameterName = parameters.get(0).getSimpleName().toString();
-    NullSavingStrategy nullSavingStrategy =
-        methodElement.getAnnotation(Insert.class).nullSavingStrategy();
+
+    Insert annotation = methodElement.getAnnotation(Insert.class);
```
Comment (tolbertam) [inline mapper-processor/src/main/java/com/datastax/oss/driver/internal/mapper/processor/dao/DaoInsertMethodGenerator.java]: This code is repeated among the next four classes, could probably be refactored similarly to how I describe in `NullSavingStrategyValidation`

**Review by tolbertam — CHANGES_REQUESTED** [review]
Looks really great, lots of good test coverage.  I had some suggestions, but :+1: to the approach.

File: mapper-processor/src/main/java/com/datastax/oss/driver/internal/mapper/processor/dao/NullSavingStrategyValidation.java (hunk 17 above)
Comment (olim7t) [reply mapper-processor/src/main/java/com/datastax/oss/driver/internal/mapper/processor/dao/NullSavingStrategyValidation.java]: Removing the default makes the attribute mandatory on the annotation.

Comment (tolbertam) [reply mapper-processor/src/main/java/com/datastax/oss/driver/internal/mapper/processor/dao/NullSavingStrategyValidation.java]: ah, right.  That didn't occur to me at the time, in any case, could we just set default to null?

File: mapper-processor/src/main/java/com/datastax/oss/driver/internal/mapper/processor/dao/DaoImplementationGenerator.java (hunk 6 above)
Comment (tomekl007) [reply mapper-processor/src/main/java/com/datastax/oss/driver/internal/mapper/processor/dao/DaoImplementationGenerator.java]: Yes, I think that this is a good idea to have a "normal" java code where possible (instead of generated). Changed (https://github.com/datastax/java-driver/pull/1246/commits/37aada0af2153961c75d4fb7714b7b42ffc56ee4)

File: mapper-processor/src/main/java/com/datastax/oss/driver/internal/mapper/processor/dao/NullSavingStrategyValidation.java (hunk 15 above)
Comment (tomekl007) [reply mapper-processor/src/main/java/com/datastax/oss/driver/internal/mapper/processor/dao/NullSavingStrategyValidation.java:123]: I was also thinking about it - I had an impression that sometimes we are favoring code readability over DRY but If that is your impression I am +1 on this as well.
Implemented in https://github.com/datastax/java-driver/pull/1246/commits/0e99daa6a413387ae8478a01b5fd88c463edbc4f

File: mapper-processor/src/main/java/com/datastax/oss/driver/internal/mapper/processor/dao/NullSavingStrategyValidation.java (hunk 17 above)
Comment (tomekl007) [reply mapper-processor/src/main/java/com/datastax/oss/driver/internal/mapper/processor/dao/NullSavingStrategyValidation.java]: If we will set it to `null` we will still need to interpret is somehow for a case where 
`@DefaultNullSavingStrategy` on the `@Dao` level is not a set AND the method level `nullSavingStrategy` property for example on `@Update` will not be set. 
In such case that null needs to become `DO_NOT_SET` or `SET_TO_NULL`. I think that it is better to have that default logic embedded within the annotation default:
`NullSavingStrategy nullSavingStrategy() default NullSavingStrategy.DO_NOT_SET;`
instead of adding another case to handle when the value is null.

File: mapper-processor/src/main/java/com/datastax/oss/driver/internal/mapper/processor/dao/DaoInsertMethodGenerator.java (hunk 21 above)
Comment (tomekl007) [reply mapper-processor/src/main/java/com/datastax/oss/driver/internal/mapper/processor/dao/DaoInsertMethodGenerator.java]: https://github.com/datastax/java-driver/pull/1246/commits/6d1e5e67a568b52db19fbc6a587288a9a593894e

File: mapper-processor/src/main/java/com/datastax/oss/driver/internal/mapper/processor/dao/NullSavingStrategyValidation.java (hunk 13 above)
Comment (tomekl007) [reply mapper-processor/src/main/java/com/datastax/oss/driver/internal/mapper/processor/dao/NullSavingStrategyValidation.java:98]: I like those names. Changed in https://github.com/datastax/java-driver/pull/1246/commits/7a40e48bbd042518390c5541b72ab67f4d7ae392

File: mapper-runtime/src/main/java/com/datastax/oss/driver/api/mapper/annotations/DefaultNullSavingStrategy.java (hunk 2 above)
Comment (tomekl007) [reply mapper-runtime/src/main/java/com/datastax/oss/driver/api/mapper/annotations/DefaultNullSavingStrategy.java:57]: I agree with you that `@NullSavingStrategy` in this context should be enough.
The `@DefaultNullSavingStrategy` name was an initial idea of @olim7t - wdyt about changing that name to `@NullSavingStrategy`

File: manual/mapper/daos/defaultnullsavingstrategy/README.md (hunk 19 above)
Comment (tomekl007) [reply manual/mapper/daos/defaultnullsavingstrategy/README.md]: The method level annotations have `NullSavingStrategy nullSavingStrategy() default NullSavingStrategy.DO_NOT_SET;` value. When the user is using it will have a default set.
Don't you think that some of our users may think that this unset default can take precedence?
I think that there is a small chance that someone may think like this so maybe it is better to have it explicitly documented?

**Review by tolbertam — APPROVED** [review]
:+1: my feedback has been addressed by recent changes.  Only question left is on the naming, but that is just a suggestion and i'm ok with either choice.

File: mapper-runtime/src/main/java/com/datastax/oss/driver/api/mapper/annotations/DefaultNullSavingStrategy.java (hunk 2 above)
Comment (olim7t) [reply mapper-runtime/src/main/java/com/datastax/oss/driver/api/mapper/annotations/DefaultNullSavingStrategy.java:57]: In theory I would agree, but then the annotation would have the same name as the enum. I find this really unfortunate given that one references the other.

File: mapper-processor/src/main/java/com/datastax/oss/driver/internal/mapper/processor/dao/NullSavingStrategyValidation.java (hunk 17 above)
Comment (olim7t) [reply mapper-processor/src/main/java/com/datastax/oss/driver/internal/mapper/processor/dao/NullSavingStrategyValidation.java]: Plus `null` is not a valid default for an annotation parameter.

**Review by olim7t — APPROVED** [review]

**Comment by olim7t** [comment]
Merged with `@DefaultNullSavingStrategy`: I felt there was a pretty strong case to not clash with the enum name.
I also took the liberty of rearranging the manual page.
